{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "##Pipeline for annotation of Nahant phage genomes\n",
    "\n",
    "No more waiting for other folks to get pipelines up and running... let's figure this out.\n",
    "\n",
    ">UBLAST of prodigal-identified proteins against several sequence databases:\n",
    ">>COG\n",
    "\n",
    ">>PFam\n",
    "\n",
    ">>ACLAME\n",
    "\n",
    ">>Camera Viral Proteins (CVP)\n",
    "\n",
    ">UBLAST run via run_ublasts.py script in ./scripts; implemented on server using ./scripts/ublast_runs.slurm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_prodigal_phage(inputfasta, out_gene, out_prot):\n",
    "    to_run=\"prodigal -i \"+inputfasta+\" -o \"+out_gene+\" -a \"+out_prot+\" -p meta\"\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_blastp(inputfasta=\"input.fasta\", output_file=\"blast.out\", database=\"databast.fasta\", evalue=\"0.001\"):\n",
    "    to_run=\"blastp -db \"+database+\" -query \"+inputfasta+\"-evalue \"+evalue+\" -outfmt 6 -out \"+output_file\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "\n",
    "def run_formatdb(fastafile, protein=\"yes\"):\n",
    "    dbtype=\"prot\"\n",
    "    if protein==\"no\":\n",
    "        dbtype=\"nucl\"\n",
    "    to_run=\"makeblastdb -in \"+fastafile+\" -dbtype \"+dbtype\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_formatudb(fastafile, databasefile=\"db.udb\", ublast_path=\"/home/sbiller/usearch7.0.1090_i86linux64\"):\n",
    "    to_run=ublast_path+\" -makeudb_ublast \"+fastafile+\" -output \"+databasefile\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_ublastp(fastafile, out_file, udb, evalue, ublast_path=\"/home/sbiller/usearch7.0.1090_i86linux64\"):\n",
    "    to_run=ublast_path+\" -ublast \"+fastafile+\" -db \"+udb+\" -evalue \"+evalue+\" -accel 0.5 -blast6out \"+out_file+\" -top_hit_only\"\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_trna_scan(input_file, output):\n",
    "    if os.path.exists(output):\n",
    "        os.remove(output)\n",
    "    args=[\"tRNAscan-SE\", \"-o\", output, \"-G\", \"-D\",\"-N\", input_file]\n",
    "    subprocess.call(args)\n",
    "    print(\"tRNA scan of \"+input_file+\" is done!\")\n",
    "    \n",
    "def find_best_hit(gene_id, dict_list):\n",
    "    evals=1\n",
    "    annotation=\"\"\n",
    "    best_hit=\"\"\n",
    "    for i in range(0, len(first_look)):\n",
    "        if gene_id in dict_list[i].keys():\n",
    "            hit=dict_list[i][gene_id]\n",
    "            #print hit[1]\n",
    "            if float(hit[1])<evals:\n",
    "                evals=float(hit[1])\n",
    "                best_hit=dict_names[i]\n",
    "                annotation=hit[-1]\n",
    "            #print(dict_names[i]+\"\\t\"+hit[1]+\"\\t\"+hit[-1])\n",
    "        \n",
    "    #print(\"best annotation for\"+gene_id+\" is from \"+best_hit+\" with e-value \"+str(evals)+\" and annotation of \"+annotation)\n",
    "    return [annotation, best_hit]\n",
    "\n",
    "#considers hits to more informative databases before less informative databases\n",
    "#dict_lists are lists of blast_dict tables and dl_names are the names of the dicts in the same order\n",
    "def find_best_hit2(gene_id, dict_list1, dl1_names, dict_list2=[], dl2_names=[]):\n",
    "    evals=1\n",
    "    annotation=\"\"\n",
    "    best_hit=\"\"\n",
    "    hits=[]\n",
    "    es=[]\n",
    "    names=[]\n",
    "    for i in range(0, len(dict_list1)):\n",
    "        if gene_id in dict_list1[i].keys():\n",
    "            hit=dict_list1[i][gene_id]\n",
    "            hits.append(hit[-1])\n",
    "            es.append(float(hit[1]))\n",
    "            names.append(dl1_names[i])\n",
    "    if len(hits)>0:\n",
    "        best_annotation=[hits[es.index(min(es))],names[es.index(min(es))]]\n",
    "    else:\n",
    "        for i in range(0, len(dict_list2)):\n",
    "            if gene_id in dict_list2[i].keys():\n",
    "                hit=dict_list2[i][gene_id]\n",
    "                hits.append(hit[-1])\n",
    "                es.append(hit[1])\n",
    "                names.append(dl2_names[i])\n",
    "        if len(hits)>0:\n",
    "            best_annotation=[hits[es.index(min(es))],names[es.index(min(es))]]\n",
    "        else:\n",
    "            best_annotation=[\"\",\"\"]\n",
    "\n",
    "    #print(\"best annotation for\"+gene_id+\" is from \"+best_hit+\" with e-value \"+str(evals)+\" and annotation of \"+annotation)\n",
    "    return best_annotation \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below: Creating DB dicts for quick access to sequence codes, annotations and OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#create DB dicts for BLAST file analysis:\n",
    "\n",
    "#ACLAME:\n",
    "aclame=open(\"./databases/DB_Info/aclame/aclame_proteins_all_0.4.tab\").readlines()\n",
    "aclame_dict={}\n",
    "\n",
    "for line in aclame[2:-1]:\n",
    "    protein=line.split(\"\\t\")[0]\n",
    "    ncbi_ann=line.split(\"\\t\")[2]\n",
    "    aclame_dict[protein]=ncbi_ann\n",
    "\n",
    "##COG needs two dbs:\n",
    "cogs=open(\"./databases/DB_Info/COG/cog2003-2014.csv\").readlines()         #all COG sequences and COG groups\n",
    "cogs2=open(\"./databases/DB_Info/COG/cognames2003-2014.tab\").readlines()   #COG group definitions/functions\n",
    "\n",
    "cog_dict={}\n",
    "cog_defs={}\n",
    "\n",
    "#dict from gi to COG:\n",
    "for line in cogs:\n",
    "    gi=line.split(\",\")[0]\n",
    "    cog=line.split(\",\")[6]\n",
    "    cog_dict[gi]=cog\n",
    "#dict from COG to function definition:\n",
    "for line in cogs2:\n",
    "    cog=line.split(\"\\t\")[0]\n",
    "    func=line.split(\"\\t\")[2]\n",
    "    cog_defs[cog]=func\n",
    "\n",
    "##Pfam needs two DBs as well:\n",
    "pfams=open(\"./databases/DB_Info/PFam/Pfam-A.titles.txt\").readlines()   #ID all pfam sequences in BLAST db\n",
    "pfams2=open(\"./databases/DB_Info/PFam/Pfam-A.clans.tsv\").readlines()   #Matches IDs to \"clans\" with functions\n",
    "pfam_dict={}\n",
    "pfam_defs={}\n",
    "\n",
    "#sequence to pfam:\n",
    "for line in pfams:\n",
    "    seq=line.split(\" \")[0].replace(\">\",\"\")\n",
    "    pfam=line.split(\" \")[2].split(\";\")[0]\n",
    "    pfam_dict[seq]=pfam\n",
    "#pfam to function:\n",
    "for line in pfams2:\n",
    "    pfam=line.split(\"\\t\")[0]\n",
    "    function=line.split(\"\\t\")[4]\n",
    "    pfam_defs[pfam]=function\n",
    "    \n",
    "##CAMERA Viral Proteins annotations are complicated; extracting definitions from sequence titles:\n",
    "cvp=open(\"./databases/DB_Info/CVP/CVP_titles.txt\").readlines()\n",
    "\n",
    "cvp_dict={}\n",
    "\n",
    "headers=[]\n",
    "\n",
    "for line in cvp:\n",
    "    defs={}\n",
    "    line=line.replace(\">\",\"\")\n",
    "    annotation=\"\"\n",
    "    info=line.split(\"/\")[1:]\n",
    "    ID=line.split(\"/\")[0].replace(\" \",\"\")\n",
    "    for i in info: \n",
    "        if \"=\" in i:\n",
    "            defs[i.split(\"=\")[0]]=i.split(\"=\")[1]\n",
    "    if \"DESCRIPTION\" in defs.keys():\n",
    "        annotation=defs[\"DESCRIPTION\"]\n",
    "    elif \"definition\" in defs.keys():\n",
    "        annotation=defs[\"definition\"]\n",
    "    \n",
    "    cvp_dict[ID]=annotation\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Now looking into the actual BLAST files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#modules for blast file processing\n",
    "def get_digits(prodfile)\n",
    "    prod=open(prodfile).readlines()\n",
    "    digits=len(str(len(prod)/2))\n",
    "    return digits\n",
    "\n",
    "\n",
    "def get_locus_tag(line, digits, phage):\n",
    "    query=line.split(\"\\t\")[0].split(\" \")[0]\n",
    "    number=query.split(\"_\")[-1]\n",
    "    z=\"0\"*(digits-len(number))\n",
    "    return \"NVP\"+phage.replace(\".\",\"\")+\"_\"+z+number\n",
    "\n",
    "#create a dict of BLAST results for each database BLAST:\n",
    "\n",
    "#ACLAME:\n",
    "def create_aclame_blast_dict(blast, aclame_dict=aclame_dict, digits=digits):\n",
    "    blast=open(blast).readlines()\n",
    "    aclame_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        annotation=aclame_dict[hit]\n",
    "        info=[hit, e, bitscore, annotation]\n",
    "        if locus_tag not in aclame_blast_dict.keys():\n",
    "            aclame_blast_dict[locus_tag]=info\n",
    "    return aclame_blast_dict\n",
    "    #print(locus_tag+\"\\t\"+hit+\"\\t\"+e+\"\\t\"+bitscore+\"\\t\"+aclame_dict[hit])\n",
    "\n",
    "#COGs:\n",
    "def create_cog_blast_dict(blast, cog_dict=cog_dict, cog_defs=cog_defs, digits=digits):\n",
    "    blast=open(blast).readlines()\n",
    "    cog_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        cog=cog_dict[(hit.split(\"|\")[1])]\n",
    "        func=cog_defs[cog].replace(\"\\n\",\"\")\n",
    "        info=[hit, e, bitscore, cog, func]\n",
    "        if locus_tag not in cog_blast_dict.keys():\n",
    "            cog_blast_dict[locus_tag]=info\n",
    "    return cog_blast_dict\n",
    "\n",
    "#PFam:\n",
    "def create_pfam_blast_dict(blast, pfam_dict=pfam_dict, pfam_defs=pfam_defs, digits=digits):\n",
    "    blast=open(blast).readlines()\n",
    "    pfam_blast_dict={}\n",
    "    \n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        pfam=pfam_dict[hit].split(\".\")[0]\n",
    "        function=pfam_defs[pfam].replace(\"\\n\",\"\")\n",
    "        info=[hit, e, bitscore, pfam, function]\n",
    "        if locus_tag not in pfam_blast_dict.keys():\n",
    "            pfam_blast_dict[locus_tag]=info\n",
    "    return pfam_blast_dict\n",
    "\n",
    "##CVP:\n",
    "\n",
    "def create_cvp_blast_dict(blast, cvp_dict=cvp_dict, digits=digits):\n",
    "    blast=open(blast).readlines()\n",
    "    cvp_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        cvp=hit.replace(\" \",\"\")\n",
    "        func=cvp_dict.get(cvp,\"\")\n",
    "        info=[hit, e, bitscore, cvp, func]\n",
    "        if locus_tag not in cvp_blast_dict.keys():\n",
    "            cvp_blast_dict[locus_tag]=info\n",
    "    return cvp_blast_dict\n",
    "\n",
    "\n",
    "def write_cds_gff3(gff, phage, out):\n",
    "    out=open(out, \"w\")\n",
    "    prod=open(gff).readlines()\n",
    "    digits=len(str(len(prod)/2))  ##for assigning a gene number with appropriate number of zeros preceding it\n",
    "    \n",
    "    OGs={\"Pfam\":pfam_blast_dict, \"COG\":cog_blast_dict}\n",
    "    first_looks=[pfam_blast_dict, aclame_blast_dict, cog_blast_dict]\n",
    "    flnames=[\"pfam\",\"aclame\",\"cog\"]\n",
    "    second_look=[cvp_blast_dict]\n",
    "    slnames=[\"cvp\"]\n",
    "    \n",
    "    SeqID=prod[0].split(\";\")[2].split(\"=\")[1].replace('\"','')\n",
    "\n",
    "    phage=gff.replace(\"gene\",\"\")\n",
    "    \n",
    "    for i in range(2,len(prod)-1,2):\n",
    "        loc=prod[i]\n",
    "\n",
    "        info=prod[i+1]\n",
    "\n",
    "        if len(loc.split())==2:\n",
    "            if \"complement\" in prod[i].split()[1]:\n",
    "                strand=\"-\"\n",
    "                start=loc.split(\"..\")[1].replace(\")\\n\",\"\")\n",
    "                stop=loc.split(\"(\")[1].split(\"..\")[0]\n",
    "            else:\n",
    "                strand=\"+\"\n",
    "                start=loc.split()[1].split(\"..\")[0]\n",
    "                stop=loc.split()[1].split(\"..\")[1].replace(\"\\n\",\"\")\n",
    "        number=info.split(\";\")[0].split(\"=\")[2].split(\"_\")[1]\n",
    "        z=\"0\"*(digits-len(number))\n",
    "        t=\"NVP\"+phage.replace(\".\",\"\")+\"_\"+z+number\n",
    "        col9=\"ID=\"+t\n",
    "        best_hits=find_best_hit2(t, dict_list1=first_looks, dl1_names=flnames, dict_list2=second_look, dl2_names=slnames)\n",
    "        Name=best_hits[0]\n",
    "        if len(Name)==0:\n",
    "            col9+=', Name=hypothetical protein'\n",
    "        else:\n",
    "            col9+=\", Name=\"+Name.replace('\"','')\n",
    "            col9+=\", note=annotation from \"+best_hits[1]\n",
    "\n",
    "        for d in OGs.keys():\n",
    "            db=OGs[d]\n",
    "            if t in db.keys():\n",
    "                col9+=', Ontology_term=\"'+d+\":\"+db[t][-2]+'\"'\n",
    "        col9+=\"\\n\"\n",
    "        out.write(SeqID+\"\\tprod\\tCDS\\t\"+start+\"\\t\"+stop+\"\\t.\\t\"+strand+\"\\t0\\t\"+col9)\n",
    "    out.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "phage=\"1.161.O.\"\n",
    "\n",
    "digits=get_digits(phage+\"gene\")\n",
    "\n",
    "#Set up BLAST dicts:\n",
    "aclame_blast_dict=create_aclame_blast_dict(blast=phage+\"vs.aclame.out\")\n",
    "cog_blast_dict=create_cog_blast_dict(blast=phage+\"vs.cogs_2003-2014.out\")\n",
    "pfam_blast_dict=create_pfam_blast_dict(blast=phage+\"vs.Pfam.out\")\n",
    "cvp_blast_dict=create_cvp_blast_dict(blast=phage+\"vs.CVP.out\")\n",
    "\n",
    "write_cds_gff3(gff, phage, phage+\"cds.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
