{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_prodigal_phage(inputfasta, out_gene, out_prot):\n",
    "    to_run=\"prodigal -i \"+inputfasta+\" -o \"+out_gene+\" -a \"+out_prot+\" -p meta\"\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_blastp(inputfasta=\"input.fasta\", output_file=\"blast.out\", database=\"databast.fasta\", evalue=\"0.001\"):\n",
    "    to_run=\"blastp -db \"+database+\" -query \"+inputfasta+\"-evalue \"+evalue+\" -outfmt 6 -out \"+output_file\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "\n",
    "def run_formatdb(fastafile, protein=\"yes\"):\n",
    "    dbtype=\"prot\"\n",
    "    if protein==\"no\":\n",
    "        dbtype=\"nucl\"\n",
    "    to_run=\"makeblastdb -in \"+fastafile+\" -dbtype \"+dbtype\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_formatudb(fastafile, databasefile=\"db.udb\", ublast_path=\"/home/sbiller/usearch7.0.1090_i86linux64\"):\n",
    "    to_run=ublast_path+\" -makeudb_ublast \"+fastafile+\" -output \"+databasefile\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_ublastp(fastafile, out_file, udb, evalue, ublast_path=\"/home/sbiller/usearch7.0.1090_i86linux64\"):\n",
    "    to_run=ublast_path+\" -ublast \"+fastafile+\" -db \"+udb+\" -evalue \"+evalue+\" -accel 0.5 -blast6out \"+out_file+\" -top_hit_only\"\n",
    "    subprocess.call(to_run.split(\" \"))\n",
    "    \n",
    "def run_trna_scan(input_file, output):\n",
    "    if os.path.exists(output):\n",
    "        os.remove(output)\n",
    "    args=[\"tRNAscan-SE\", \"-o\", output, \"-G\", \"-D\",\"-N\", input_file]\n",
    "    subprocess.call(args)\n",
    "    print(\"tRNA scan of \"+input_file+\" is done!\")\n",
    "    \n",
    "def find_best_hit(gene_id, dict_list):\n",
    "    evals=1\n",
    "    annotation=\"\"\n",
    "    best_hit=\"\"\n",
    "    for i in range(0, len(first_look)):\n",
    "        if gene_id in dict_list[i].keys():\n",
    "            hit=dict_list[i][gene_id]\n",
    "            #print hit[1]\n",
    "            if float(hit[1])<evals:\n",
    "                evals=float(hit[1])\n",
    "                best_hit=dict_names[i]\n",
    "                annotation=hit[-1]\n",
    "            #print(dict_names[i]+\"\\t\"+hit[1]+\"\\t\"+hit[-1])\n",
    "        \n",
    "    #print(\"best annotation for\"+gene_id+\" is from \"+best_hit+\" with e-value \"+str(evals)+\" and annotation of \"+annotation)\n",
    "    return [annotation, best_hit]\n",
    "\n",
    "#considers hits to more informative databases before less informative databases\n",
    "#dict_lists are lists of blast_dict tables and dl_names are the names of the dicts in the same order\n",
    "def find_best_hit2(gene_id, dict_list1, dl1_names, dict_list2=[], dl2_names=[]):\n",
    "    evals=1\n",
    "    annotation=\"\"\n",
    "    best_hit=\"\"\n",
    "    hits=[]\n",
    "    es=[]\n",
    "    names=[]\n",
    "    for i in range(0, len(dict_list1)):\n",
    "        if gene_id in dict_list1[i].keys():\n",
    "            hit=dict_list1[i][gene_id]\n",
    "            hits.append(hit[-1])\n",
    "            es.append(float(hit[1]))\n",
    "            names.append(dl1_names[i])\n",
    "    if len(hits)>0:\n",
    "        best_annotation=[hits[es.index(min(es))],names[es.index(min(es))]]\n",
    "    else:\n",
    "        for i in range(0, len(dict_list2)):\n",
    "            if gene_id in dict_list2[i].keys():\n",
    "                hit=dict_list2[i][gene_id]\n",
    "                hits.append(hit[-1])\n",
    "                es.append(hit[1])\n",
    "                names.append(dl2_names[i])\n",
    "        if len(hits)>0:\n",
    "            best_annotation=[hits[es.index(min(es))],names[es.index(min(es))]]\n",
    "        else:\n",
    "            best_annotation=[\"\",\"\"]\n",
    "\n",
    "    #print(\"best annotation for\"+gene_id+\" is from \"+best_hit+\" with e-value \"+str(evals)+\" and annotation of \"+annotation)\n",
    "    return best_annotation \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "format_blast_databases.py                     100%  272     0.3KB/s   00:00    \n",
      "formatdb1.slurm                               100%  455     0.4KB/s   00:00    \n"
     ]
    }
   ],
   "source": [
    "!scp ./databases/scripts/* jbrown@eofe4.mit.edu:/nobackup1/jbrown/annotation/databases/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.161.O.vs.aclame.out                         100%   32KB  31.6KB/s   00:00    \n",
      "1.161.O.vs.cogs_2003-2014.out                 100%  271KB 270.7KB/s   00:00    \n",
      "1.161.O.vs.CVP.out                            100%  260KB 259.9KB/s   00:00    \n",
      "1.161.O.vs.kegg.out                           100%    0     0.0KB/s   00:00    \n",
      "1.161.O.vs.Pfam.out                           100%  771KB 771.3KB/s   00:00    \n"
     ]
    }
   ],
   "source": [
    "!scp jbrown@eofe4.mit.edu:/nobackup1/jbrown/annotation/blasts/*/1.161.O.*.out ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.161.O.gene                                  100%   54KB  54.0KB/s   00:00    \n"
     ]
    }
   ],
   "source": [
    "!scp jbrown@eofe4.mit.edu:/nobackup1/jbrown/annotation/genes/1.161.O.* ./"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below processing a prodigal gff file into a sequin table as described here: http://www.ncbi.nlm.nih.gov/genbank/genomesubmit-examples#fig2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#create DB dicts for BLAST file analysis:\n",
    "\n",
    "#ACLAME:\n",
    "aclame=open(\"./databases/DB_Info/aclame/aclame_proteins_all_0.4.tab\").readlines()\n",
    "aclame_dict={}\n",
    "\n",
    "for line in aclame[2:-1]:\n",
    "    protein=line.split(\"\\t\")[0]\n",
    "    ncbi_ann=line.split(\"\\t\")[2]\n",
    "    aclame_dict[protein]=ncbi_ann\n",
    "\n",
    "##COG needs two dbs:\n",
    "cogs=open(\"./databases/DB_Info/COG/cog2003-2014.csv\").readlines()         #all COG sequences and COG groups\n",
    "cogs2=open(\"./databases/DB_Info/COG/cognames2003-2014.tab\").readlines()   #COG group definitions/functions\n",
    "\n",
    "cog_dict={}\n",
    "cog_defs={}\n",
    "\n",
    "#dict from gi to COG:\n",
    "for line in cogs:\n",
    "    gi=line.split(\",\")[0]\n",
    "    cog=line.split(\",\")[6]\n",
    "    cog_dict[gi]=cog\n",
    "#dict from COG to function definition:\n",
    "for line in cogs2:\n",
    "    cog=line.split(\"\\t\")[0]\n",
    "    func=line.split(\"\\t\")[2]\n",
    "    cog_defs[cog]=func\n",
    "\n",
    "##Pfam needs two DBs as well:\n",
    "pfams=open(\"./databases/DB_Info/PFam/Pfam-A.titles.txt\").readlines()   #ID all pfam sequences in BLAST db\n",
    "pfams2=open(\"./databases/DB_Info/PFam/Pfam-A.clans.tsv\").readlines()   #Matches IDs to \"clans\" with functions\n",
    "pfam_dict={}\n",
    "pfam_defs={}\n",
    "\n",
    "#sequence to pfam:\n",
    "for line in pfams:\n",
    "    seq=line.split(\" \")[0].replace(\">\",\"\")\n",
    "    pfam=line.split(\" \")[2].split(\";\")[0]\n",
    "    pfam_dict[seq]=pfam\n",
    "#pfam to function:\n",
    "for line in pfams2:\n",
    "    pfam=line.split(\"\\t\")[0]\n",
    "    function=line.split(\"\\t\")[4]\n",
    "    pfam_defs[pfam]=function\n",
    "    \n",
    "##CAMERA Viral Proteins annotations are complicated; extracting definitions from sequence titles:\n",
    "cvp=open(\"./databases/DB_Info/CVP/CVP_titles.txt\").readlines()\n",
    "\n",
    "cvp_dict={}\n",
    "\n",
    "headers=[]\n",
    "\n",
    "for line in cvp:\n",
    "    defs={}\n",
    "    line=line.replace(\">\",\"\")\n",
    "    annotation=\"\"\n",
    "    info=line.split(\"/\")[1:]\n",
    "    ID=line.split(\"/\")[0].replace(\" \",\"\")\n",
    "    for i in info: \n",
    "        if \"=\" in i:\n",
    "            defs[i.split(\"=\")[0]]=i.split(\"=\")[1]\n",
    "    if \"DESCRIPTION\" in defs.keys():\n",
    "        annotation=defs[\"DESCRIPTION\"]\n",
    "    elif \"definition\" in defs.keys():\n",
    "        annotation=defs[\"definition\"]\n",
    "    \n",
    "    cvp_dict[ID]=annotation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gff=\"1.161.O.gene\"\n",
    "\n",
    "prod=open(gff).readlines()\n",
    "\n",
    "digits=len(str(len(prod)/2))  ##for assigning a gene number with appropriate number of zeros preceding it\n",
    "\n",
    "phage=gff.replace(\"gene\",\"\")\n",
    "for i in range(2,len(prod)-1,2):\n",
    "    loc=prod[i]\n",
    "    info=prod[i+1]\n",
    "    feature1=\"gene\"\n",
    "    feature2=\"CDS\"\n",
    "    if len(loc.split())==2:\n",
    "        if \"complement\" in prod[i].split()[1]:\n",
    "            start=loc.split(\"..\")[1].replace(\")\\n\",\"\")\n",
    "            stop=loc.split(\"(\")[1].split(\"..\")[0]\n",
    "        else:\n",
    "            start=loc.split()[1].split(\"..\")[0]\n",
    "            stop=loc.split()[1].split(\"..\")[1].replace(\"\\n\",\"\")\n",
    "    number=info.split(\";\")[0].split(\"=\")[2].split(\"_\")[1]\n",
    "\n",
    "    z=\"0\"*(digits-len(number))\n",
    "    \n",
    "    locus_tag=\"nvp_\"+phage+\"_\"+z+number\n",
    "    #print(start+\"\\t\"+stop+\"\\t\"+feature1+\"\\n\\t\\t\\tlocus_tag\"+\"\\t\"+locus_tag+\"\\n\")\n",
    "    #print(start+\"\\t\"+stop+\"\\t\"+feature2+\"\\n\\t\\t\\tprotein_id\"+\"\\t\"+locus_tag+\"\\n\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Now looking into the actual BLAST files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#modules for blast file processing\n",
    "prodfile=\"1.161.O.gene\"\n",
    "prod=open(prodfile).readlines()\n",
    "digits=len(str(len(prod)/2))\n",
    "\n",
    "\n",
    "def get_locus_tag(line, digits, phage):\n",
    "    query=line.split(\"\\t\")[0].split(\" \")[0]\n",
    "    number=query.split(\"_\")[-1]\n",
    "    z=\"0\"*(digits-len(number))\n",
    "    return \"NVP\"+phage.replace(\".\",\"\")+\"_\"+z+number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#create a dict of BLAST results\n",
    "blast=\"1.161.O.vsaclame.out\"\n",
    "\n",
    "def create_aclame_blast_dict(blast, aclame_dict=aclame_dict):\n",
    "    blast=open(blast).readlines()\n",
    "    aclame_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        annotation=aclame_dict[hit]\n",
    "        info=[hit, e, bitscore, annotation]\n",
    "        if locus_tag not in aclame_blast_dict.keys():\n",
    "            aclame_blast_dict[locus_tag]=info\n",
    "    return aclame_blast_dict\n",
    "    #print(locus_tag+\"\\t\"+hit+\"\\t\"+e+\"\\t\"+bitscore+\"\\t\"+aclame_dict[hit])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#COGs:\n",
    "def create_cog_blast_dict(blast, cog_dict=cog_dict, cog_defs=cog_defs):\n",
    "    blast=open(\"./1.161.O.vscogs_2003-2014.out\").readlines()\n",
    "    cog_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        cog=cog_dict[(hit.split(\"|\")[1])]\n",
    "        func=cog_defs[cog].replace(\"\\n\",\"\")\n",
    "        info=[hit, e, bitscore, cog, func]\n",
    "        if locus_tag not in cog_blast_dict.keys():\n",
    "            cog_blast_dict[locus_tag]=info\n",
    "    return cog_blast_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#now Pfam results:\n",
    "def create_pfam_blast_dict(blast, pfam_dict=pfam_dict, pfam_defs=pfam_defs):\n",
    "    blast=open(blast).readlines()\n",
    "    pfam_blast_dict={}\n",
    "    \n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        pfam=pfam_dict[hit].split(\".\")[0]\n",
    "        function=pfam_defs[pfam].replace(\"\\n\",\"\")\n",
    "        info=[hit, e, bitscore, pfam, function]\n",
    "        if locus_tag not in pfam_blast_dict.keys():\n",
    "            pfam_blast_dict[locus_tag]=info\n",
    "    return pfam_blast_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##CVP BLAST results\n",
    "\n",
    "def create_cvp_blast_dict(blast, cvp_dict=cvp_dict)\n",
    "    blast=open(\"./1.161.O.vs.CVP.out\").readlines()\n",
    "    cvp_blast_dict={}\n",
    "\n",
    "    for line in blast:\n",
    "        locus_tag=get_locus_tag(line, digits, phage)\n",
    "        hit=line.split(\"\\t\")[1]\n",
    "        e=line.split(\"\\t\")[-2]\n",
    "        bitscore=line.split(\"\\t\")[-1].replace(\"\\n\",\"\")\n",
    "        cvp=hit.replace(\" \",\"\")\n",
    "        func=cvp_dict.get(cvp,\"\")\n",
    "        info=[hit, e, bitscore, cvp, func]\n",
    "        if locus_tag not in cvp_blast_dict.keys():\n",
    "            cvp_blast_dict[locus_tag]=info\n",
    "    return cvp_blast_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Setup list of BLAST dicts:\n",
    "first_looks=[pfam_blast_dict, aclame_blast_dict, cog_blast_dict]\n",
    "flnames=[\"pfam\",\"aclame\",\"cog\"]\n",
    "second_look=[cvp_blast_dict]\n",
    "slnames=[\"cvp\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NVP1161O_001\t['', '']\n",
      "NVP1161O_002\t['Terminase-like family', 'pfam']\n",
      "NVP1161O_003\t['C-5 cytosine-specific DNA methylase', 'pfam']\n",
      "NVP1161O_004\t['', '']\n",
      "NVP1161O_005\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_006\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_007\t['', '']\n",
      "NVP1161O_008\t['T5orf172 domain', 'pfam']\n",
      "NVP1161O_009\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_010\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_011\t['Putative phage serine protease XkdF', 'pfam']\n",
      "NVP1161O_012\t['T5orf172 domain', 'pfam']\n",
      "NVP1161O_013\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_014\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_015\t['Phage major capsid protein E', 'pfam']\n",
      "NVP1161O_016\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_017\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_018\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_019\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_020\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_021\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_022\t['', '']\n",
      "NVP1161O_023\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_024\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_025\t['', '']\n",
      "NVP1161O_026\t['', '']\n",
      "NVP1161O_027\t['\"hypothetical protein ViPhICP1_gp117 [Vibrio phage ICP1].\" ', 'cvp']\n",
      "NVP1161O_028\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_029\t['', '']\n",
      "NVP1161O_030\t['', '']\n",
      "NVP1161O_031\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_032\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_033\t['', '']\n",
      "NVP1161O_034\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_035\t['', '']\n",
      "NVP1161O_036\t['Helix-turn-helix', 'pfam']\n",
      "NVP1161O_037\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_038\t['', '']\n",
      "NVP1161O_039\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_040\t['', '']\n",
      "NVP1161O_041\t['', '']\n",
      "NVP1161O_042\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_043\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_044\t['Protein of unknown function (DUF559)', 'pfam']\n",
      "NVP1161O_045\t['', '']\n",
      "NVP1161O_046\t['', '']\n",
      "NVP1161O_047\t['CRISPR associated protein Cas1', 'pfam']\n",
      "NVP1161O_048\t['CRISPR/Cas system-associated endonuclease/helicase Cas3', 'cog']\n",
      "NVP1161O_049\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_050\t['CRISPR-associated protein (Cas_Csy2)', 'pfam']\n",
      "NVP1161O_051\t['CRISPR-associated protein (Cas_Csy3)', 'pfam']\n",
      "NVP1161O_052\t['CRISPR-associated protein (Cas_Csy4)', 'pfam']\n",
      "NVP1161O_053\t['Phage conserved protein', 'aclame']\n",
      "NVP1161O_054\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_055\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_056\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_057\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_058\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_059\t['T5orf172 domain', 'pfam']\n",
      "NVP1161O_060\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_061\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_062\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_063\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_064\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_065\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_066\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_067\t['Phage conserved protein', 'aclame']\n",
      "NVP1161O_068\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_069\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_070\t['putative tail fiber protein', 'aclame']\n",
      "NVP1161O_071\t['', '']\n",
      "NVP1161O_072\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_073\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_074\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_075\t['', '']\n",
      "NVP1161O_076\t['', '']\n",
      "NVP1161O_077\t['', '']\n",
      "NVP1161O_078\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_079\t['', '']\n",
      "NVP1161O_080\t['', '']\n",
      "NVP1161O_081\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_082\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_083\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_084\t['', '']\n",
      "NVP1161O_085\t['Putative DNA polymerase', 'aclame']\n",
      "NVP1161O_086\t['\"putative DNA-binding protein 2 [Bacillus phage B4].\" ', 'cvp']\n",
      "NVP1161O_087\t['AAA domain', 'pfam']\n",
      "NVP1161O_088\t['GIY-YIG catalytic domain', 'pfam']\n",
      "NVP1161O_089\t['DNA primase/helicase', 'aclame']\n",
      "NVP1161O_090\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_091\t['', '']\n",
      "NVP1161O_092\t['', '']\n",
      "NVP1161O_093\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_094\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_095\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_096\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_097\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_098\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_099\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_100\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_101\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_102\t['', '']\n",
      "NVP1161O_103\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_104\t['', '']\n",
      "NVP1161O_105\t['Resolvase, N terminal domain', 'pfam']\n",
      "NVP1161O_106\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_107\t['', '']\n",
      "NVP1161O_108\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_109\t['', '']\n",
      "NVP1161O_110\t['UV-endonuclease UvdE', 'pfam']\n",
      "NVP1161O_111\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_112\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_113\t['', '']\n",
      "NVP1161O_114\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_115\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_116\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_117\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_118\t['TM2 domain', 'pfam']\n",
      "NVP1161O_119\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_120\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_121\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_122\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_123\t['', '']\n",
      "NVP1161O_124\t['', '']\n",
      "NVP1161O_125\t['', '']\n",
      "NVP1161O_126\t['', '']\n",
      "NVP1161O_127\t['', '']\n",
      "NVP1161O_128\t['', '']\n",
      "NVP1161O_129\t['T5orf172 domain', 'pfam']\n",
      "NVP1161O_130\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_131\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_132\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_133\t['', '']\n",
      "NVP1161O_134\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_135\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_136\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_137\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_138\t['', '']\n",
      "NVP1161O_139\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_140\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_141\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_142\t['', '']\n",
      "NVP1161O_143\t['DNA ligase', 'aclame']\n",
      "NVP1161O_144\t['', '']\n",
      "NVP1161O_145\t['', '']\n",
      "NVP1161O_146\t['Clp protease', 'pfam']\n",
      "NVP1161O_147\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_148\t['', '']\n",
      "NVP1161O_149\t['D12 class N6 adenine-specific DNA methyltransferase', 'pfam']\n",
      "NVP1161O_150\t['PhoH-like protein', 'pfam']\n",
      "NVP1161O_151\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_152\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_153\t['', '']\n",
      "NVP1161O_154\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_155\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_156\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_157\t['Ribonucleotide reductase beta subunit, ferritin-like domain', 'cog']\n",
      "NVP1161O_158\t['similar to other intron-encoded site specific endonucleases', 'aclame']\n",
      "NVP1161O_159\t['Ribonucleotide reductase, small chain', 'pfam']\n",
      "NVP1161O_160\t['Ribonucleotide reductase alpha subunit', 'cog']\n",
      "NVP1161O_161\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_162\t['', '']\n",
      "NVP1161O_163\t['', '']\n",
      "NVP1161O_164\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_165\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_166\t['D12 class N6 adenine-specific DNA methyltransferase', 'pfam']\n",
      "NVP1161O_167\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_168\t['Dihydrofolate reductase', 'pfam']\n",
      "NVP1161O_169\t['Thymidylate synthase', 'pfam']\n",
      "NVP1161O_170\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_171\t['', '']\n",
      "NVP1161O_172\t['Calcineurin-like phosphoesterase', 'pfam']\n",
      "NVP1161O_173\t['', '']\n",
      "NVP1161O_174\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_175\t['', '']\n",
      "NVP1161O_176\t['', '']\n",
      "NVP1161O_177\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_178\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_179\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_180\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_181\t['', '']\n",
      "NVP1161O_182\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_183\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_184\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_185\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_186\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_187\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_188\t['', '']\n",
      "NVP1161O_189\t['\"hypothetical protein ViPhICP1_gp166 [Vibrio phage ICP1].\" ', 'cvp']\n",
      "NVP1161O_190\t['Putative exonuclease, RdgC', 'pfam']\n",
      "NVP1161O_191\t['SPFH domain / Band 7 family', 'pfam']\n",
      "NVP1161O_192\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_193\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_194\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_195\t[\"5' nucleotidase, deoxy (Pyrimidine), cytosolic type C protein (NT5C)\", 'pfam']\n",
      "NVP1161O_196\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_197\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_198\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_199\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_200\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_201\t['', '']\n",
      "NVP1161O_202\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_203\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_204\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_205\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_206\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_207\t['', '']\n",
      "NVP1161O_208\t['Protein chain release factor B', 'cog']\n",
      "NVP1161O_209\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_210\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_211\t['Phosphopantetheine attachment site', 'pfam']\n",
      "NVP1161O_212\t['', '']\n",
      "NVP1161O_213\t['AAA domain', 'pfam']\n",
      "NVP1161O_214\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_215\t['hypothetical protein', 'aclame']\n",
      "NVP1161O_216\t['', '']\n",
      "NVP1161O_217\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_218\t['', '']\n",
      "NVP1161O_219\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_220\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_221\t['', '']\n",
      "NVP1161O_222\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_223\t['\"hypothetical protein\" ', 'cvp']\n",
      "NVP1161O_224\t['putative phage-related protein', 'aclame']\n",
      "NVP1161O_225\t['', '']\n",
      "NVP1161O_226\t['', '']\n",
      "NVP1161O_227\t['\"predicted protein\" ', 'cvp']\n",
      "NVP1161O_228\t['hypothetical protein', 'aclame']\n"
     ]
    }
   ],
   "source": [
    "gff=\"1.161.O.gene\"\n",
    "prod=open(gff).readlines()\n",
    "digits=len(str(len(prod)/2))  ##for assigning a gene number with appropriate number of zeros preceding it\n",
    "\n",
    "phage=gff.replace(\"gene\",\"\")\n",
    "for i in range(2,len(prod)-1,2):\n",
    "    loc=prod[i]\n",
    "    info=prod[i+1]\n",
    "    feature1=\"gene\"\n",
    "    feature2=\"CDS\"\n",
    "    if len(loc.split())==2:\n",
    "        if \"complement\" in prod[i].split()[1]:\n",
    "            start=loc.split(\"..\")[1].replace(\")\\n\",\"\")\n",
    "            stop=loc.split(\"(\")[1].split(\"..\")[0]\n",
    "        else:\n",
    "            start=loc.split()[1].split(\"..\")[0]\n",
    "            stop=loc.split()[1].split(\"..\")[1].replace(\"\\n\",\"\")\n",
    "    number=info.split(\";\")[0].split(\"=\")[2].split(\"_\")[1]\n",
    "    z=\"0\"*(digits-len(number))\n",
    "    locus_tag=\"NVP\"+phage.replace(\".\",\"\")+\"_\"+z+number\n",
    "    \n",
    "    print(locus_tag+\"\\t\"+str(find_best_hit2(locus_tag, dict_list1=first_looks, dl1_names=flnames, dict_list2=second_look, dl2_names=slnames)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gff=\"1.161.O.gene\"\n",
    "prod=open(gff).readlines()\n",
    "digits=len(str(len(prod)/2))  ##for assigning a gene number with appropriate number of zeros preceding it\n",
    "\n",
    "tags=[]\n",
    "\n",
    "phage=gff.replace(\"gene\",\"\")\n",
    "for i in range(2,len(prod)-1,2):\n",
    "    loc=prod[i]\n",
    "    info=prod[i+1]\n",
    "    feature1=\"gene\"\n",
    "    feature2=\"CDS\"\n",
    "    if len(loc.split())==2:\n",
    "        if \"complement\" in prod[i].split()[1]:\n",
    "            start=loc.split(\"..\")[1].replace(\")\\n\",\"\")\n",
    "            stop=loc.split(\"(\")[1].split(\"..\")[0]\n",
    "        else:\n",
    "            start=loc.split()[1].split(\"..\")[0]\n",
    "            stop=loc.split()[1].split(\"..\")[1].replace(\"\\n\",\"\")\n",
    "    number=info.split(\";\")[0].split(\"=\")[2].split(\"_\")[1]\n",
    "    z=\"0\"*(digits-len(number))\n",
    "    tags.append(\"NVP\"+phage.replace(\".\",\"\")+\"_\"+z+number)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "OGs={\"Pfam\":pfam_blast_dict, \"COG\":cog_blast_dict}\n",
    "\n",
    "out=open(\"annotation_practice.txt\",\"w\")\n",
    "\n",
    "for t in tags:\n",
    "    out.write(\"ID=\"+t)\n",
    "    best_hits=find_best_hit2(t, dict_list1=first_looks, dl1_names=flnames, dict_list2=second_look, dl2_names=slnames)\n",
    "    Name=best_hits[0]\n",
    "    if len(Name)==0:\n",
    "        out.write(', Name=hypothetical protein')\n",
    "    else:\n",
    "        out.write(\", Name=\"+Name.replace('\"',''))\n",
    "        out.write(\", note=annotation from \"+best_hits[1])\n",
    "    \n",
    "    for d in OGs.keys():\n",
    "        db=OGs[d]\n",
    "        if t in db.keys():\n",
    "            out.write(', Ontology_term=\"'+d+\":\"+db[t][-2]+'\"')\n",
    "    out.write(\"\\n\")\n",
    "out.close()\n",
    "!open annotation_practice.txt -a textedit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "aclame_hit=aclame_blast_dict.get(locus_tag,\"no aclame\")\n",
    "    cog_hit=cog_blast_dict.get(locus_tag, \"no cog\")\n",
    "    pfam_hit=pfam_blast_dict.get(locus_tag, \"no pfam\")\n",
    "    cvp_hit=cvp_blast_dict.get(locus_tag, \"no cvp\")\n",
    "    best_hit=find_best_hit2(locus_tag, dict_list1=first_looks, dl1_names=flnames, dict_list2=second_look, dl2_names=slnames)\n",
    "    \n",
    "    print(aclame_hit)\n",
    "    print(cog_hit)\n",
    "    print(pfam_hit)\n",
    "    print(cvp_hit)\n",
    "    \n",
    "    \n",
    "    print(locus_tag+\"\\t\"+str(find_best_hit2(locus_tag, dict_list1=first_looks, dl1_names=flnames, dict_list2=second_look, dl2_names=slnames)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
